xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWiki.java

```diff
@@ -2435,17 +2470,61 @@ public String getSkinFile(String filename, boolean forceSkinAction, XWikiContext
 
         // If all else fails, use the default base skin, even if the URLs could be invalid.
         URL url;
+
         if (forceSkinAction) {
             url = urlf.createSkinURL(filename, "skins", getDefaultBaseSkin(context), context);
         } else {
             url = urlf.createSkinURL(filename, getDefaultBaseSkin(context), context);
         }
-        return urlf.getURL(url, context);
+        queryParameters.putAll(getResourceURLParameters(url));
+        return urlf.getURL(url, context, queryParameters);
+    }
+
+    public String getSkinFile(String filename, boolean forceSkinAction, XWikiContext context)
+    {
+        return getSkinFile(filename, forceSkinAction, context, new LinkedHashMap<>());
+    }
+
+    private Map<String, String> getResourceURLParameters(Resource<?> resource)
+    {
+        try {
+            URL resourceUrl = getResource(resource.getPath());
+            return getResourceURLParameters(resourceUrl);
+        } catch (MalformedURLException e) {
+            LOGGER.debug("Error while getting URL for resource filepath [{}]", resource.getPath(), e);
+            return Collections.singletonMap(CACHE_VERSION, getVersion());
+        }
+    }
+
+    private Map<String, String> getResourceURLParameters(String resourceFilePath)
+    {
+        try {
+            URL resourceUrl = getResource(resourceFilePath);
+            return getResourceURLParameters(resourceUrl);
+        } catch (MalformedURLException e) {
+            LOGGER.debug("Error while getting URL for resource filepath [{}]", resourceFilePath, e);
+            return Collections.singletonMap(CACHE_VERSION, getVersion());
+        }
+    }
+
+    private Map<String, String> getResourceURLParameters(URL resourceUrl)
+    {
+        Map<String, String> parameters = new LinkedHashMap<>();
+
+        try {
+            Path resourcePath = Paths.get(resourceUrl.toURI());
+            FileTime lastModifiedTime = Files.getLastModifiedTime(resourcePath);
+            parameters.put(CACHE_VERSION, String.valueOf(lastModifiedTime.toMillis()));
```

Comment:
Personally I would have preferred using the XWiki version for performance reasons. Note that if you change some files, you could have also changed the XWiki version for example with a prefix "10.8.3-patch1". It's logical since it means "patching" xwiki. Right now I think you need to restart XWiki for the version change to take effect but either that could be acceptable or we could have imagine a way to set/override the XWiki version in the XWiki Admin UI.

I don't know how much perfs are going to be impacted but this means a ton of IO more for large wikis. And this is is unnecessary on the very large majority of cases (you don't patch xwiki every second!).